### DISCUSSION TITLE
Dismissing layers and cleanup

### DISCUSSION THREAD

**Discussion by alexandremjacques** [unpoly/unpoly#584] [General]
Hi,

I'm having some issues integrating TinyMCE with Unpoly (inside layers, TinyMCE doesn't reinitialize when the layer is closed then reopened).

I kinda solved the issue by using the `up:layer:dismiss` event to find and remove the editor. It's a bit generic though since I have to check if the layer has an editor to remove in the first place (in the case of closing other layers that don't have an editor).

Just wondering if there's a better way to do this.

**Comment by triskweline** [comment]
> Just wondering if there's a better way to do this.

The canonical way to initialize third-party libraries is a [compiler](https://unpoly.com/up.compiler).

Compilers can return a [destructor function](https://unpoly.com/up.compiler#destructor) that is called when the element is destroyed.. When a layer is closed, all fragments in that layer are destroyed and their destructor functions are called.

I found this TinyMCE example in a project:

```js
up.compiler('[wysiwyg-editor]', (element) => {
  let editor

  TinyMCE.init({
    target: element,
    init_instance_callback: (instance) => editor = instance,
  })

  return () => { editor?.destroy() }
})
```

**Comment by alexandremjacques** [comment]
That's great! Worked like a charm. Thanks!
